Dockerfile

```diff
@@ -70,6 +70,16 @@ COPY --from=wheel-cache /cache /cache
 RUN pip install ${PIP_CACHE_DIRECTIVE} --upgrade pip==19.0.1 && \
     pip install ${PIP_CACHE_DIRECTIVE} --no-use-pep517 -e ".[${EXTRAS}]"
 
+# Install necessary NPM dependencies - also taking into account only changes in package.json
+COPY airflow/www/ /opt/airflow/airflow/www/
+
+WORKDIR /opt/airflow/airflow/www
+
+RUN npm ci \
```

Comment:
AHAH! That command is _hard_ to find in the docs. ✨

Reply:
True :). Install seems to be quite obvious candidate to install locked versions of packages. In fact it seems it used to work like that before NPM 5.1. This is one of the highest voted Slack questions/answers I stumbled upon recently:
https://stackoverflow.com/questions/45022048/why-does-npm-install-rewrite-package-lock-json

Reply:
I think I looked at that about 2 weeks ago and that "update 3" wasn't there - 5.7.0 was only just released :D

Reply:
We are living on a bleeding edge ;)

Reply:
Oh BAH!

> If a node_modules is already present, it will be automatically removed before npm ci begins its install.

this means (for my other project, not airflow at least) that `npm ci` takes far longer than just `npm install` as it seems to bust the cache :(

(They took a short-cut in the implementation I feel. I'd like a mode that checks the current versions, and doesn't wipe the node_modules if everything is up-to-date)

Reply:
In our case (of Docker Builds) it's not a problem. It's all but guaranteed that there will not be node_modules when we reach that instruction in Docker. For CI build, it's always from the scratch. For local builds the node_modules dir is excluded in .dockerignore so it is not even sent as part of the context to the Docker engine performing the build.

Now when you wrote it however, I actually think that I should add another layer for that (COPY packages-json.lock -> npm ci -> COPY www -> npm run prod). Then 'npm ci' will only be run  when packages-lock.json changes - which should be very rare case.

Reply:
```
WORKDIR /opt/airflow/airflow/www

# Install necessary NPM dependencies - also taking into account only changes in package.json
COPY --chown=airflow:airflow airflow/www/package-lock.json /opt/airflow/airflow/www/package-lock.json

RUN npm ci

# Install necessary NPM dependencies - also taking into account only changes in package.json
COPY --chown=airflow:airflow airflow/www/ /opt/airflow/airflow/www/

RUN npm run prod

WORKDIR /opt/airflow

```

Reply:
Actually it gets even better:  with setting custom prefix in Docker -> this way node_modules will be installed and used from /cache/node_modules - not messing with the code of the application. So even if we mount the sources from local directory we will keep node_modules installed via Docker (and you need both package and package-lock to make 'npm ci' works :)). Just checked.

Reply:
I will take a look at that still. As mentioned in another thread, node_modules are now generated in the source code. That causes a number of problems when the sources are mounted (the node_modules might be generated either in docker or from an external npm which might be different version. 

I will take a look at the possibility of moving the node_modules permanently out of the sources tree inside docker (not in the host). This will keep the local development workflow (and IDE integration for example) untouched - node_modules will still be in the sources. Since the node_modules are already exclueded from the Docker context via .dockerignore this is fine for CI execution, but moving node_modules out will help in case the sources are mounted to inside the container.

Reply:
OK. I changed it in the latest push in #4932 I fixed it in the way that npm install && npm run prod are run inside the container when you enter it - in case when the "node_modules" and "dist" folders do not exist (i.e. when the sources are mounted from the host and npm was never run on the host). I think that should cover the most common use case for development and this way we avoid the case that modules/prepared .js files are missing. The user will still be able to run "npm install && run prod inside the container" manually (or even outside!).